Add checking if the new node is not the same as original mode in the setmode function

vava24680 opened the pull request:
I added the checking of whether the new mode and the original mode are the same or not in the `setmode` function. If both of the modes are the same, it's regarded as a normal behavior.
I have checked the original RPi.GPIO source code, there is similar checking mechanism in the `setmode` function.

vava24680 committed:
Add checking if the new node is not the same as original mode

Signed-off-by: Yu-Hao Chang <9203700@gmail.com>

theyoyojo commented:
Hey, thanks for the suggestion.

I took a look at the original RPi.GPIO code and while you are correct that there is a condition that resembles this, it seems to me that the condition is logically redundant in the use cases actually supported by RPi.GPIO.

Do you know of a use case where this condition would be relevant, and if so could you add a small test for it in the `tests/test_gpio.py` file?

vava24680 commented:
Hi @theyoyojo, could you give me some time to answer ? I have something more important to do now, sorry for the delay.

vava24680 committed:
Add tests on the repeated mode setup

Signed-off-by: Yu-Hao Chang <9203700@gmail.com>

vava24680 commented:
@theyoyojo I found this use case when I use the [Adafruit_CircuitPython_DHT](https://github.com/adafruit/Adafruit_CircuitPython_DHT) library to get the temperature data from a DHT11 sensor. In that package, it uses `setmode` function twice.

I think since this package tries to replace to the origin RPi.GPIO, so it's reasonable to maintain the original behavior as the original RPi.GPIO, that's why I added this check in the `setmode` function.

I already added a test for this change.

Thank you!

theyoyojo commented:
Hi,

Thank you for your test case. If we cannot find a use case for the extra logical condition, then I will at the very least merge your test case.

However, I took a look at https://github.com/adafruit/Adafruit_CircuitPython_DHT and I was unable to find any references to RPi.GPIO in that library. I noticed that it depends on https://github.com/adafruit/Adafruit_Blinka, so I checked that codebase as well, and though it does indeed use RPi.GPIO, I could not find an instance where `setmode()` is called twice in a context where it would be executed on a Raspberry Pi. All I found was this: https://github.com/adafruit/Adafruit_Blinka/blob/1c7e43354667e4268f8d806d9ee1bd3d41272310/src/adafruit_blinka/microcontroller/bcm283x/pin.py#L4

Can you point me to the specific codepath you found that requires this logical condition?

vava24680 commented:
1. In the [adafruit_dht.py](https://github.com/adafruit/Adafruit_CircuitPython_DHT/blob/master/adafruit_dht.py), at line 33, it imports `DigitalInOut`, `Pull`, `Direction` from the [digitalio.py](https://github.com/adafruit/Adafruit_Blinka/blob/master/src/digitalio.py), and In the `digitalio.py`, it further imports the pin module. In the [corresponding pin module](https://github.com/adafruit/Adafruit_Blinka/blob/master/src/adafruit_blinka/microcontroller/bcm283x/pin.py), the `setmode` is called first time.
2. Also in the  [adafruit_dht.py](https://github.com/adafruit/Adafruit_CircuitPython_DHT/blob/master/adafruit_dht.py), at line 37, it imports the `PulseIn` from the [pluseio.py](https://github.com/adafruit/Adafruit_Blinka/blob/master/src/pulseio.py) and in the `pluseio.py`, the [`PulseIn.py`](https://github.com/adafruit/Adafruit_Blinka/blob/master/src/adafruit_blinka/microcontroller/bcm283x/pulseio/PulseIn.py) and [`PWMOut.py`](https://github.com/adafruit/Adafruit_Blinka/blob/master/src/adafruit_blinka/microcontroller/bcm283x/pulseio/PWMOut.py) are imported, the `setmode` is called second time in the `PWMOut.py`. **That's why I need this logical condition.**

theyoyojo commented:
I am still having trouble reproducing this on my RPi 3 B v1.2, but I will try this out on my RPi 4B+ as well. Once I can reproduce the issue and confirm that your patch fixes it, I will merge the pull request.

As noted in #28, we don't have a reliable way to detect RPi version info in a distro-generic manner, and as you can see by the current values in the RPI_INFO dictionary, the data is hardcoded for the RPi 3 B v1.2 at the moment, though we plan to change this in the near future. One main goal of this project is to provide support for the RPi.GPIO API on as many distributions as possible.
